src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectHandler.cs

```diff
@@ -350,36 +347,35 @@ public OpenIdConnectHandler(HttpClient backchannel)
                 {
                     return null;
                 }
+                message = messageReceivedContext.ProtocolMessage;
 
-                var properties = new AuthenticationProperties();
-
-                // if state is missing, just log it
-                if (string.IsNullOrEmpty(message.State))
+                // if any of the error fields are set, throw error null
+                if (!string.IsNullOrEmpty(message.Error))
                 {
-                    Logger.LogWarning(Resources.OIDCH_0004_MessageStateIsNullOrEmpty);
+                    Logger.LogError(Resources.OIDCH_0006_MessageContainsError, message.Error, message.ErrorDescription ?? "ErrorDecription null", message.ErrorUri ?? "ErrorUri null");
+                    throw new OpenIdConnectProtocolException(string.Format(CultureInfo.InvariantCulture, Resources.OIDCH_0006_MessageContainsError, message.Error, message.ErrorDescription ?? "ErrorDecription null", message.ErrorUri ?? "ErrorUri null"));
                 }
-                else
-                {
-                    // if state exists and we failed to 'unprotect' this is not a message we should process.
-                    properties = Options.StateDataFormat.Unprotect(Uri.UnescapeDataString(message.State));
-                    if (properties == null)
-                    {
-                        Logger.LogError(Resources.OIDCH_0005_MessageStateIsInvalid);
-                        return null;
-                    }
 
-                    string userstate = null;
-                    properties.Items.TryGetValue(OpenIdConnectDefaults.UserstatePropertiesKey, out userstate);
-                    message.State = userstate;
+                // Fail if state is missing, it's required for the correlation id.
+                if (string.IsNullOrEmpty(message.State))
+                {
+                    Logger.LogError(Resources.OIDCH_0004_MessageStateIsNullOrEmpty);
+                    return null;
                 }
 
-                // if any of the error fields are set, throw error null
-                if (!string.IsNullOrEmpty(message.Error))
+                // if state exists and we failed to 'unprotect' this is not a message we should process.
+                var properties = Options.StateDataFormat.Unprotect(Uri.UnescapeDataString(message.State));
```

Comment:
Is `Uri.UnescapeDataString` needed here? The OAuth2 middleware doesn't use it: https://github.com/aspnet/Security/blob/dev/src/Microsoft.AspNet.Authentication.OAuth/OAuthHandler.cs#L100